activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb

```diff
@@ -48,7 +48,7 @@ def perform_query(raw_connection, sql, binds, type_casted_binds, prepare:, notif
             # made since we established the connection
             raw_connection.query_options[:database_timezone] = default_timezone
 
-            result = if !prepared_statements || binds.nil? || binds.empty?
+            result = if (binds.nil? || binds.empty?) && !prepare
               ActiveSupport::Dependencies.interlock.permit_concurrent_loads do
```

Comment:
I also note in passing that this is the only adapter (and prior to #53702, only in the prepare=true case) that currently includes `permit_concurrent_loads`. Additional stack frame ugliness notwithstanding, perhaps it should live in `raw_execute`?

Reply:
Yep, agreed. My only thought for this PR would be to remove it, as that matches the other adapters and the currently-more-common MySQL-without-prepare cases.

Reply:
(but it also is more correct for it to be there, so 🤷🏻‍♂️😅)